### ISSUE TITLE
Logged in backend user overrules frontend user session when retreiving basket

### ISSUE THREAD

**Issue by Physikbuddha** [aimeos/aimeos-typo3#226]
This issue is not easily explained in two sentences.

While testing our website shop running Aimeos, I created a fresh, new frontend user and logged myself in. We are not using the frontend user creation feature of Aimeos, but have our own registration process on top.

I put an item in the basket and proceeded to the checkout flow.
I noticed that for some unknown reason during the "Enter Address" step, none of the radio buttons to choose an address was selected, so I tried to click Next without checking anything to see what will happen.
I was expecting an error message, but the checkout process just went on to the next step (Payment method).

This behaviour is not really wanted, but let's keep that aside for a second.
When I got to the "Summary" step, I was expecting an empty "Invoice address", since I didn't check any.
However, the summary step was showing the full address of the user that was logged in previously.

To make things short, I did some research and debugging into this.
It turned out, that the same basket from the previous user was recycled for the newly created user.

Why did this happen?
Well, it's because I was still logged in into the TYPO3 backend with the same backend admin user.

To fetch the frontend user's basket, a `$key` is generated using `$key = $token . '-' . $sitecode . '-' . $language . '-' . $currency . '-' . $type;`.
`$token` is fetched from the current logged in user's session.

The code responsible for fetching the token resides in [aimeos/aimeos-typo3/Classes/Base/Context.php->addToken()](https://github.com/aimeos/aimeos-typo3/blob/c2d9c19bba15170c6473599dc4283c9a7dd1137e/Classes/Base/Context.php#L298).

However, before fetching the frontend user's ID, a check to the current session is made whether it already contains a `token`.
When there's an active backend login, the token of that backend user will be returned.

This behaviour is unexpected in my eyes. I would expect an Aimeos basket being tied to the current frontend user, not the backend one.

**Comment by aimeos** [comment]
Guess, the main problem is that the context prefers TYPO3 backend sessions over frontend sessions:
https://github.com/aimeos/aimeos-typo3/blob/c2d9c19bba15170c6473599dc4283c9a7dd1137e/Classes/Base/Context.php#L272-L275

If I remember correctly, this was done to avoid security issues.

**Comment by nvindice** [comment]
@aimeos There is one more problem with this behaviour: when returning from an external payment provider, the backend session cookie is not available because of a strict SameSite setting. So the session is "lost" and the payment fails. However, when reloading the page, the session becomes available again.

Long story short: I'm not able to buy something using an external payment provider, while logged in at the backend.

Any idea?

**Comment by aimeos** [comment]
That should be no real problem because it's only a testing issue, isn't it?
Like said, this is a security related design decision.

**Comment by nvindice** [comment]
That's correct; the problem only arises during testing (though it is very annoying there).

To be honest, I don't understand the reasoning behind the security concerns - IMHO, this actually increases the attack surface (leaked cookies, XSS, deliberate context separation in Typo3 is circumvented, the principle of least privilege is violated). On top of that, it makes debugging incredibly difficult.

But if it’s a deliberate design decision, it’s none of my business. Thanks for the feedback.

**Comment by aimeos** [comment]
Use two browser windows, one in private mode. Then, your problem is gone.
The security issue behind is that frontend cookies shouldn't take precedence over backend cookies.
